### PR TITLE
fix: clearing timers race condition

### PR DESCRIPTION
**Pull request by bartlomieju**
Fixes https://github.com/denoland/deno/issues/7599

<s>Still need to come up with a test</s>

### CODE REVIEW

**Comment by magurotuna** [comment]
Great work!

File: cli/tests/integration_tests.rs [hunk 1]
```diff
@@ -1371,6 +1371,11 @@ itest!(_004_set_timeout {
   output: "004_set_timeout.ts.out",
 });
 
+itest!(_004_set_timeout_clear {
+  args: "run --quiet --reload 004_set_timeout_clear.ts",
+  output: "004_set_timeout_clear.ts.out",
+});
+
```
Comment (magurotuna) [inline cli/tests/integration_tests.rs:1378]: I think this test could pass without the change in this PR. We would be happier if we could confirm that it finishes immediately, not take 10 seconds. Is this difficult to check?

Comment (bartlomieju) [reply cli/tests/integration_tests.rs:1378]: Good point, I'll rewrite the test

Comment (magurotuna) [reply cli/tests/integration_tests.rs]: Cool :+1:

File: cli/tests/set_timeout_clear.ts [hunk 2]
```diff
@@ -0,0 +1,5 @@
+const handle = setTimeout(() => {
+  console.log("timeout finish");
+}, 10000);
+clearTimeout(handle);
+console.log("finish");
```
Comment (ry) [inline cli/tests/set_timeout_clear.ts:5]: Surely this is currently working?

File: cli/ops/timers.rs [hunk 3]
```diff
@@ -56,12 +59,16 @@ impl GlobalTimer {
...
+      .then(|_| futures::future::ok(()))
+      .boxed_local();
+    self.future = Some(fut);
   }
 }
 
 pub fn init(rt: &mut deno_core::JsRuntime) {
   super::reg_json_sync(rt, "op_global_timer_stop", op_global_timer_stop);
+  super::reg_json_sync(rt, "op_global_timer_start", op_global_timer_start);
   super::reg_json_async(rt, "op_global_timer", op_global_timer);
```
Comment (ry) [inline cli/ops/timers.rs:72]: It's not clear to me what these three ops do. Can you give a description? (maybe put it in a comment?)

File: cli/tests/integration_tests.rs [hunk 4]
```diff
@@ -1480,6 +1480,29 @@ itest!(deno_test_only {
...
+  use std::time::Duration;
+  use std::time::Instant;
+
+  let mut p = util::deno_cmd()
+    .current_dir(util::tests_path())
+    .arg("run")
+    .arg("--reload")
+    .arg("set_timeout_clear.ts")
+    .spawn()
+    .unwrap();
```
Comment (ry) [inline cli/tests/integration_tests.rs:1496]: Would be cool the read the source code from stdin, so you didn't have to create a separate file `set_timeout_clear.ts`

File: cli/ops/timers.rs [hunk 5]
```diff
@@ -5,7 +5,7 @@
 //! As an optimization, we want to avoid an expensive calls into rust for every
 //! setTimeout in JavaScript. Thus in //js/timers.ts a data structure is
 //! implemented that calls into Rust for only the smallest timeout.  Thus we
-//! only need to be able to start and cancel a single timer (or Delay, as Tokio
+//! only need to be able to start, cancel and awwait a single timer (or Delay, as Tokio
```
Comment (magurotuna) [inline cli/ops/timers.rs:8]: > awwait

typo?

File: cli/tests/integration_tests.rs (hunk 4 above)
Comment (ry) [reply cli/tests/integration_tests.rs:1504]: nice work

**Review by ry — APPROVED** [review]
LGTM - nice test, nice fix.

**Pull request merged**